File: plugins/bitcoin/bitcoin.5m.go
//go:build ignore

package main

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
)

func crossbar(args ...string) string {
	cmd := exec.Command("crossbar", args...)
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func formatNumber(n float64) string {
	str := strconv.FormatFloat(n, 'f', 0, 64)
	result := ""
	for i, c := range str {
		if i > 0 && (len(str)-i)%3 == 0 {
			result += ","
		}
		result += string(c)
	}
	return result
}

func main() {
	url := "https://api.coinbase.com/v2/prices/BTC-USD/spot"
	response := crossbar("--web", url, "--json")

	if response == "" {
		fmt.Println("₿ Error")
		fmt.Println("---")
		fmt.Println("Failed to fetch price")
		return
	}

	var data map[string]interface{}
	if err := json.Unmarshal([]byte(response), &data); err != nil {
		fmt.Println("₿ Parse Error")
		return
	}

	price := "--"
	formatted := "--"
	if d, ok := data["data"].(map[string]interface{}); ok {
		if amount, ok := d["amount"].(string); ok {
			price = amount
			if p, err := strconv.ParseFloat(amount, 64); err == nil {
				formatted = formatNumber(p)
			}
		}
	}

	fmt.Printf("₿ $%s\n", formatted)
	fmt.Println("---")
	fmt.Printf("BTC/USD: $%s\n", price)
	fmt.Println("Source: Coinbase")
	fmt.Println("---")
	fmt.Println("Refresh | refresh=true")
}

File: plugins/bitcoin/bitcoin.5m.rs
//! Bitcoin Price Plugin - Uses Crossbar API for HTTP requests

use std::process::Command;
use serde_json::Value;

fn crossbar(args: &[&str]) -> Option<String> {
    let output = Command::new("crossbar")
        .args(args)
        .output()
        .ok()?;
    
    if output.status.success() {
        Some(String::from_utf8_lossy(&output.stdout).trim().to_string())
    } else {
        None
    }
}

fn format_number(n: f64) -> String {
    let s = format!("{:.0}", n);
    let chars: Vec<char> = s.chars().collect();
    let mut result = String::new();
    for (i, c) in chars.iter().enumerate() {
        if i > 0 && (chars.len() - i) % 3 == 0 {
            result.push(',');
        }
        result.push(*c);
    }
    result
}

fn main() {
    let url = "https://api.coinbase.com/v2/prices/BTC-USD/spot";
    let response = crossbar(&["--web", url, "--json"]);

    if response.is_none() {
        println!("₿ Error");
        println!("---");
        println!("Failed to fetch price");
        return;
    }

    let response = response.unwrap();
    match serde_json::from_str::<Value>(&response) {
        Ok(data) => {
            let price = data["data"]["amount"]
                .as_str()
                .unwrap_or("--");
            
            let formatted = price.parse::<f64>()
                .map(|p| format_number(p))
                .unwrap_or_else(|_| price.to_string());

            println!("₿ ${}", formatted);
            println!("---");
            println!("BTC/USD: ${}", price);
            println!("Source: Coinbase");
        }
        Err(_) => println!("₿ Parse Error"),
    }

    println!("---");
    println!("Refresh | refresh=true");
}
